Control statement go (#92)
JankariTech/blog: src/assets/JourneyWithGo-ABlogSeries/goControlFlow.md

---
title: Control Statements - Control Your Go
authorName: Prarup Gurung (प्रारुप गुरुङ्ग)
authorAvatar: https://www.jankaritech.com/images/2021/12/29/p1110938-crop-colour-800.jpg
authorLink: https://github.com/grgprarup
createdAt: Dec 13, 2022
tags: branching statement, if-else, switch, looping statement, control flow statement
banner: https://camo.githubusercontent.com/2b507540e2681c1a25698f246b9dca69c30548ed66a7323075b0224cbb1bf058/68747470733a2f2f676f6c616e672e6f72672f646f632f676f706865722f6669766579656172732e6a7067
seriesTitle: Journey With Go - A Blog Series
episode: 2
---

**[गो](https://go.dev "Go in Nepali")** **"Journey With Go - A Blog Series"** about the basics of Go `Go Basic`, advanced concepts in Go `Go Beyond Basics`, testing with go `Godog as Test Framework` and many more.

> This is the second installment of the series. You are here that means you have already read the first blog of the series. If you haven't read the first blog, please read it first. It is available [here](https://blog.jankaritech.com/#/blog/Journey%20With%20Go%20-%20A%20Blog%20Series/Go%20Basics%20-%20The%20Starting%20Point "here"). In this blog, we will learn about the `Control Flow Statements in Go`. **So, let's get started.**

Reviewer: ```suggestion
This is the second installment of the series. You are here which means you have already read the first blog of the series. If you haven't read the first blog, please read it first. It is available [here](https://blog.jankaritech.com/#/blog/Journey%20With%20Go%20-%20A%20Blog%20Series/Go%20Basics%20-%20The%20Starting%20Point "here"). In this blog, we will learn about the `Control Flow Statements in Go`. **So, let's get started.**
```

## **Control Flow Statements**

These statements are used to control the flow of the program. In other words, they control the execution of the program. There are two types of control flow statements in Go.

1. Branching Statements
2. Looping Statements

### **A. Branching Statements**

> Branching statements divides the program into different branches on the basis of given condition. Some blocks of code might execute and some might not. In Go, we have `if-else` and `switch` to control the flow of program.

Reviewer: ```suggestion
Branching statements divide the program into different branches on the basis of a given condition. Some blocks of code might execute and some might not. In Go, we have `if-else` and `switch` to control the flow of the program.
```

#### 1. if-else

If the condition is true, it executes the code inside the `if` block otherwise (i.e. the condition is false), it executes the code inside the `else` block. There are different forms of `if-else` pairs such as, `if`, `if-else`, `nested if-else`, and `else-if ladder`.

##### _a. if_

> This can be use if you want to execute the code only if the condition is true. The code inside of the `if` block gets executed if the condition is true. The syntax of `if` is as follows.

Reviewer 2: ```suggestion
This can be used if you want to execute the code only if the condition is true. The code inside of the `if` block gets executed if the condition is true. The syntax of `if` is as follows.
```

> This can be use if you want to execute the code only if the condition is true. The code inside of the `if` block gets executed if the condition is true. The syntax of `if` is as follows.

Reviewer: I think both sentences mean the same thing so one can be removed, can be either one of those. 
```suggestion
This can be used to execute the code only if the condition is true. The syntax of `if` is as follows.
```

```go
if condition {
    // code to execute if the condition is true
}
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/if-example.png "If Example")

Output:

```go
You are eligible to receive a Citizenship
```

##### _b. if-else_

> When you want to execute codes either for true or false case then `if-else` can be used. `if-else` executes the code inside of the `if` block if the condition is true otherwise, it executes the code inside of the `else` block. The syntax of `if-else` is as follows.

Reviewer: ```suggestion
When you want to execute codes either for true or false cases then `if-else` can be used. `if-else` executes the code inside of the `if` block if the condition is true otherwise, it executes the code inside of the `else` block. The syntax of `if-else` is as follows.
```

```go
if condition {
    // code to execute if the condition is true
} else {
    // code to execute if the condition is false
}
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/if-else-example.png "If-Else Example")

Output:

```go
You are eligible to vote.
```

##### _c. nested if-else_

> If the if-else statement is inside another if-else statement, it is called a nested if-else statement. We can use nested if-else statements to check another conditions if previous condition is met. We have different variation of nested if-else statements. We will discuss only one of them.

Reviewer: ```suggestion
If the if-else statement is inside another if-else statement, it is called a nested if-else statement. We can use nested if-else statements to check other conditions if the previous condition is met. We have different variations of nested if-else statements. We will discuss only one of them.
```

    // code to execute if the condition is true
    if condition {
        // code to execute if the condition is true
    } else {
        // code to execute if the condition is false
    }
} else {
    if condition {
        // code to execute if the condition is true
    } else {
        // code to execute if the condition is false
    }
}
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/nested-if-example.png "Nested If-Else Example")

Output:

```go
The greatest number is: 566
```

##### _d. else-if ladder_

Else-if ladder is a series of if-else statements. It is used when we have multiple conditions to check, when the previous condition is not met. The syntax of `else-if ladder` is as follows.

```go
if condition {
    // code to execute if the condition is true
} else if condition {
    // code to execute if the condition is true
} else if condition {
    // code to execute if the condition is true
} else {
    // code to execute if none of the conditions are true
}
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/else-if-ladder-example.png "Else-If Ladder Example")

Output:

```go
It is Negative number
```

### 2. Switch

> `switch` allows to execute one block of codes among several cases based on the value of a given expression. It can be used as alternative to the `if-else` statement, when you have lots of condition to be checked. If none of the cases match with the expression, the code inside of the `default` block gets executed. The syntax of `switch` is as follows.

Reviewer: ```suggestion
`switch` allows us to execute one block of code among several cases based on the value of a given expression. It can be used as an alternative to the `if-else` statement when you have lots of conditions to be checked. If none of the cases match with the expression, the code inside of the `default` block gets executed. The syntax of `switch` is as follows.
```

```go
switch expression {

>     case value1:
>         // code to execute if the expression match with value1
>     case value2:
>         // code to execute if the expression match with value2
>     case value3:
>         // code to execute if the expression match with value3
>     default:
>         // code to execute if none of the cases match with the expression
> }

Reviewer: Should we also add a break statement here or is it not needed?
```suggestion
    case value1:
        // code to execute if the expression match with value1
        break
    case value2:
        // code to execute if the expression match with value2
         break
    case value3:
        // code to execute if the expression match with value
         break
    default:
        // code to execute if none of the cases match with the expression
}
```

Author: Thank You :+1: Break is not needed for Go. I have added more explanation for this.

```
Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/switch-example.png "Switch Example")

Output:

```go
Tuesday
```

### **B. Looping Statements**

> If we want to execute block of codes repeatedly, we can use looping statements. The blocks of codes executes repeatedly until the condition is met. In Go, we have two ways to use loop the codes. They are `for` and `range`.

Reviewer: ```suggestion
If we want to execute the block of codes repeatedly, we can use looping statements. The block of codes executes repeatedly until the condition is met. In Go, we have two ways to use a loop in the codes. They are `for` and `range`.
```

#### 1. for

> `for` loop allows program to execute a block of code repeatedly until the condition is true. The syntax of `for` is as follows.

Reviewer: ```suggestion
`for` loop allows a program to execute a block of code repeatedly if the condition is true. The syntax of `for` is as follows.
```

Output:

```go
1
2
3
4
5
```

We don't have separate `while` loop in Go like other programming languages. We can use `for` loop as a `while` loop as follows.

```go
for condition {
    // code to execute if the condition is true
}
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/while-example.png "While Example")

Output:

```go
1
2
3
4
5
```

#### 2. range

`range` in Go language is used to iterate over arrays, slices, maps, and strings. The `range` is used along with the `for` loop. The syntax of `range` is as follows.

```go
for index, value := range collection {
    // code to execute until every element of the collection is iterated
}
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/range-example.png "Range Example")

Output:

```go
Fruit at index 0 is Apple
Fruit at index 1 is Mango
Fruit at index 2 is Grape
Fruit at index 3 is Lichi
Fruit at index 4 is Strawberry
```
### **C. Other Statements**

Beside, if, switch, for, and range, Go language also provides some other statements. They are `break`, `continue`, `defer`, and `panic`.
#### 1. Break Statement

> `break` statements terminates the loop or switch statement and transfers execution to the code immediately after the loop or switch. The keyword `break` is used.

Reviewer: ```suggestion
`break` statement terminates the loop or switch statement and transfers execution to the code immediately after the loop or switch. The keyword `break` is used.
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/break-example.png "Break Example")

In the above example, the `break` statement terminates the `for` loop when the value of `i` is equal to 5. So, the output will be as follows.

Output:

```go
0
1
2
3
4
Done
```

#### 2. Continue Statement

> `continue` statements skip the current iteration of the loop and continues with the next iteration. The keyword `continue` is used.

Reviewer: ```suggestion
`continue` statement skips the current iteration of the loop and continues with the next iteration. The keyword `continue` is used.
```

```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/panic-example.png "Panic Example")

In the above example, the `panic` statement is raised by the user. The program terminates as panic is raised. The output will be as follows.

Output:

```go
panic: Something went wrong

goroutine 1 [running]:
main.main()
	/home/prarup/JankariTech/playground/go-code/panic.go:7 +0x27
exit status 2
```

#### 5. Recover Statement

`recover` statement recovers from the `panic` state and resumes normal execution of the program if possible. The `recover` statement is used to catch the `panic` statement. The syntax of `recover` is as follows.

```go
recover()
```

Example:

![image](/src/assets/JourneyWithGo-ABlogSeries/images/recover-example.png "Recover Example")

In the above example, the `recover` statement is used to catch the `panic` statement. The output will be as follows.

Output:

```go
Something went wrong
```

## What we have learned so far

- Control Flow
  - Branching Statements
    - if-else
    - switch
  - Looping Statements
    - for
    - range
  - Break Statement
  - Continue Statement
  - Defer Statement
  - Panic Statement
  - Recover Statement

## Conclusion

> In this blog, we have learned about the control flow in Go language. We have learned about the branching statements, and looping statements. We have also learned about the `break`, `continue`, `defer`, `panic`, and `recover` statements. In he next blog, we will learn about the functions, Arrays, Strings, Structures in Go language. *Keep learning and keep practicing*. **Stay tuned!!!**.

Reviewer: ```suggestion
In this blog, we have learned about the control flow in Go language. We have learned about the branching statements, and looping statements. We have also learned about the `break`, `continue`, `defer`, `panic`, and `recover` statements. In the next blog, we will learn about the functions, Arrays, Strings, Structures in Go language. *Keep learning and keep practicing*. **Stay tuned!!!**.
```

---
title: Control Statements - Control Your Go
authorName: Prarup Gurung (प्रारुप गुरुङ्ग)
authorAvatar: https://www.jankaritech.com/images/2021/12/29/p1110938-crop-colour-800.jpg
authorLink: https://github.com/grgprarup

> createdAt: Dec 13, 2022

Reviewer: Dec 13, 2022? 🤔 
```suggestion
createdAt: Dec 13, 2022
```

Author: :smile: mmm that's the date when I started to write. I am also confused about which date we need to put (e.g. started date, first git commit date, PR approved date, PR merged date)? any suggestions?

Reviewer: I don't think it's that big of an issue, but I would expect the date to be in 2023 somewhere around when PR was created and reviewed/approved(somewhere around the end of march or the beginning of April?)

    case value2:
        // code to execute if the expression match with value2
    case value3:
        // code to execute if the expression match with value3
    default:
        // code to execute if none of the cases match with the expression
}
```

![Switch Flowchart](/src/assets/JourneyWithGo-ABlogSeries/images/switch.svg "Switch Flowchart")

The `break` keyword is not required, as the `switch` statement automatically breaks the execution of the code as soon as the first case is matched.

But, we can use `fallthrough` keyword to execute the code inside of the next case. The syntax of `fallthrough` is as follows.

```go
switch expression {
    case value1:
        // code to execute if the expression match with value1
        fallthrough
    case value2:
        // code to execute if the expression match with value2
    case value3:
        // code to execute if the expression match with value3
    default:
        // code to execute if none of the cases match with the expression
}
```

![Switch Fallthrough Flowchart](/src/assets/JourneyWithGo-ABlogSeries/images/switch-fallthrough.svg "Switch Fallthrough Flowchart")

**Example:**

> ![Switch Example](/src/assets/JourneyWithGo-ABlogSeries/images/switch-example.png "Switch Example")

Author: Thank you. :smile:

        // code to execute if the condition is false
    }
}
```

![Nested If-Else Flowchart](/src/assets/JourneyWithGo-ABlogSeries/images/nested-if.svg "Nested If-Else Flowchart")

**Example:**

![Nested If-Else Example](/src/assets/JourneyWithGo-ABlogSeries/images/nested-if-example.png "Nested If-Else Example")

**Output:**

```go
The greatest number is: 566
```

##### _d. else-if ladder_

Else-if ladder is a series of if-else statements. It is used when we have multiple conditions to check, when the previous condition is not met. The syntax of `else-if ladder` is as follows.

```go
if condition {
    // code to execute if the condition is true
} else if condition {
    // code to execute if the condition is true
} else if condition {
    // code to execute if the condition is true
} else {
    // code to execute if none of the conditions are true
}
```

![Else-If Ladder Flowchart](/src/assets/JourneyWithGo-ABlogSeries/images/else-if-ladder.svg "Else-If Ladder Flowchart")

**Example:**

![Else-If Ladder Example](/src/assets/JourneyWithGo-ABlogSeries/images/else-if-ladder-example.png "Else-If Ladder Example")

**Output:**

```go
It is Negative number
```

### 2. Switch

> `switch` allows us to execute one block of code among several cases based on the value of a given expression. It can be used as an alternative to the `if-else` statement, when you have lots of conditions to be checked. If none of the cases match with the expression, the code inside of the `default` block gets executed. The syntax of `switch` is as follows.

Author: I have clarified it by adding an alternative example in the switch section. :smile:
